lib/unread/reader.rb

```diff
@@ -26,13 +26,20 @@ def forget_memoized_read_mark_global
       end
 
       def have_read?(readable)
-        if self.respond_to?(:read_mark_id)
+        if self.respond_to?(:read_mark_id) and read_mark_id_belongs_to?(readable)
           # For use with scope "with_read_marks_for"
           !self.read_mark_id.nil?
         else
           !self.class.have_not_read(readable).exists?(self.id)
         end
       end
+
+      private
+
+      def read_mark_id_belongs_to?(readable)
+        self.read_mark_readable_type == readable.class.name and
```

Comment:
Small bug: Must be compared to `readable.class.base_class.name`. Also, I would prefer to use `&&` over `and` and `||` over `or`.